File: frontend/test/ai_summary_test.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

import 'package:nowcast/src/api/api_client.dart';
import 'package:nowcast/src/models/ai_summary.dart';
import 'package:nowcast/src/widgets/ai_summary_section.dart';

/// A minimal raw current-conditions payload (see API §5.1).
Map<String, dynamic> _conditionsJson() {
  return <String, dynamic>{
    'temperature': <String, dynamic>{'degrees': 33.0, 'unit': 'CELSIUS'},
    'feelsLikeTemperature': <String, dynamic>{
      'degrees': 38.0,
      'unit': 'CELSIUS',
    },
    'heatIndex': <String, dynamic>{'degrees': 41.0, 'unit': 'CELSIUS'},
    'relativeHumidity': 70,
    'weatherCondition': <String, dynamic>{
      'description': <String, dynamic>{'text': 'Sunny', 'languageCode': 'en'},
      'type': 'CLEAR',
    },
  };
}

/// A raw hourly forecast payload.
Map<String, dynamic> _hourlyJson() {
  return <String, dynamic>{
    'forecastHours': <Map<String, dynamic>>[
      <String, dynamic>{
        'weatherCondition': <String, dynamic>{
          'description': <String, dynamic>{
            'text': 'Sunny',
            'languageCode': 'en',
          },
          'type': 'CLEAR',
        },
      },
    ],
  };
}

/// A raw daily forecast payload.
Map<String, dynamic> _dailyJson() {
  return <String, dynamic>{
    'forecastDays': <Map<String, dynamic>>[
      <String, dynamic>{
        'daytimeForecast': <String, dynamic>{
          'weatherCondition': <String, dynamic>{
            'description': <String, dynamic>{
              'text': 'Partly sunny',
              'languageCode': 'en',
            },
            'type': 'PARTLY_CLOUDY',
          },
        },
      },
    ],
  };
}

/// Builds an [ApiClient] whose `ai/summary` POST returns [summaryJson].
ApiClient _buildApi(Map<String, dynamic> summaryJson) {
  final mock = MockClient((http.Request request) async {
    if (request.url.path.endsWith('/ai/summary')) {
      final body = jsonDecode(request.body) as Map<String, dynamic>;
      // Verify the section forwards the raw payloads.
      expect(body['currentConditions'], _conditionsJson());
      expect(body['hourlyForecast'], _hourlyJson());
      expect(body['dailyForecast'], _dailyJson());
      return http.Response(
        jsonEncode(summaryJson),
        200,
        headers: <String, String>{'content-type': 'application/json'},
      );
    }
    return http.Response('{"message":"Not found."}', 404);
  });
  return ApiClient(client: mock, baseUrl: 'http://example.com/api');
}

/// Builds an [ApiClient] whose `ai/summary` POST fails with [status].
ApiClient _buildFailingApi({int status = 502}) {
  final mock = MockClient((http.Request request) async {
    if (request.url.path.endsWith('/ai/summary')) {
      return http.Response('{"message":"Upstream error."}', status);
    }
    return http.Response('{"message":"Not found."}', 404);
  });
  return ApiClient(client: mock, baseUrl: 'http://example.com/api');
}

Widget _wrap({required ApiClient api}) {
  return MaterialApp(
    home: Scaffold(
      body: AiSummarySection(
        currentConditions: _conditionsJson(),
        hourlyForecast: _hourlyJson(),
        dailyForecast: _dailyJson(),
        api: api,
      ),
    ),
  );
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('AiSummary.fromJson', () {
    test('parses all three fields', () {
      final summary = AiSummary.fromJson(<String, dynamic>{
        'summary': 'Hot and humid.',
        'healthAdvice': 'Stay hydrated.',
        'articleUrl': 'https://example.com/article',
      });
      expect(summary.summary, 'Hot and humid.');
      expect(summary.healthAdvice, 'Stay hydrated.');
      expect(summary.articleUrl, 'https://example.com/article');
    });

    test('defaults missing or non-string fields to empty strings', () {
      final summary = AiSummary.fromJson(<String, dynamic>{
        'summary': 42,
        'healthAdvice': null,
      });
      expect(summary.summary, '');
      expect(summary.healthAdvice, '');
      expect(summary.articleUrl, '');
    });
  });

  group('AiSummarySection', () {
    testWidgets('renders summary, advice, and article link on success',
        (tester) async {
      final api = _buildApi(<String, dynamic>{
        'summary': 'Hot and humid with a heat index of 41°C.',
        'healthAdvice': 'Limit strenuous activity and drink water.',
        'articleUrl': 'https://www.cdc.gov/heat/index.html',
      });

      await tester.pumpWidget(_wrap(api: api));
      expect(find.text('AI summary'), findsOneWidget);
      // Loading indicator while the request is in flight.
      expect(find.byType(CircularProgressIndicator), findsOneWidget);

      await tester.pumpAndSettle();

      expect(
        find.text('Hot and humid with a heat index of 41°C.'),
        findsOneWidget,
      );
      expect(
        find.text('Limit strenuous activity and drink water.'),
        findsOneWidget,
      );
      expect(find.text('Recommended reading'), findsOneWidget);
      expect(find.byType(CircularProgressIndicator), findsNothing);
    });

    testWidgets('shows a fallback message when the response is empty',
        (tester) async {
      final api = _buildApi(<String, dynamic>{});

      await tester.pumpWidget(_wrap(api: api));
      await tester.pumpAndSettle();

      expect(
        find.text('The AI summary is unavailable right now.'),
        findsOneWidget,
      );
      expect(find.text('Recommended reading'), findsNothing);
    });

    testWidgets('shows a compact error with retry on failure', (tester) async {
      final api = _buildFailingApi(status: 502);

      await tester.pumpWidget(_wrap(api: api));
      await tester.pumpAndSettle();

      expect(find.text('Upstream error.'), findsOneWidget);
      expect(find.text('Try again'), findsOneWidget);
      // Health reminder / other content is unaffected (no full-page error).
      expect(find.byType(CircularProgressIndicator), findsNothing);
    });

    testWidgets('retry recovers after a transient failure', (tester) async {
      var calls = 0;
      final mock = MockClient((http.Request request) async {
        if (!request.url.path.endsWith('/ai/summary')) {
          return http.Response('{"message":"Not found."}', 404);
        }
        calls++;
        if (calls == 1) {
          return http.Response('{"message":"Upstream error."}', 502);
        }
        return http.Response(
          jsonEncode(<String, dynamic>{
            'summary': 'Clearing skies.',
            'healthAdvice': 'Enjoy the day.',
            'articleUrl': 'https://example.com/article',
          }),
          200,
          headers: <String, String>{'content-type': 'application/json'},
        );
      });

      await tester.pumpWidget(
        _wrap(api: ApiClient(client: mock, baseUrl: 'http://example.com/api')),
      );
      await tester.pumpAndSettle();
      expect(find.text('Try again'), findsOneWidget);

      await tester.tap(find.text('Try again'));
      await tester.pumpAndSettle();

      expect(find.text('Clearing skies.'), findsOneWidget);
      expect(find.text('Enjoy the day.'), findsOneWidget);
      expect(find.text('Recommended reading'), findsOneWidget);
      expect(find.text('Try again'), findsNothing);
    });
  });
}

File: backend/tests/Feature/AiSummaryTest.php
<?php

use App\Exceptions\OpenRouterApiException;
use App\Services\OpenRouterService;
use Illuminate\Http\Client\ConnectionException;
use Illuminate\Support\Facades\Http;

beforeEach(function (): void {
    config()->set('services.openrouter.api_key', 'test-openrouter-key');
    config()->set('services.openrouter.model', 'deepseek/deepseek-v4-flash-0731:online');
});

test('ai summary returns the model analysis', function (): void {
    Http::fake([
        'openrouter.ai/*' => Http::response([
            'choices' => [
                [
                    'message' => [
                        'content' => json_encode([
                            'summary' => 'Hot and humid with a heat index of 41°C.',
                            'healthAdvice' => 'Limit strenuous outdoor activity and stay hydrated.',
                            'articleUrl' => 'https://www.cdc.gov/heat-health/about/index.html',
                        ]),
                    ],
                ],
            ],
        ], 200),
    ]);

    $this->postJson('/api/ai/summary', [
        'currentConditions' => [
            'temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS'],
            'heatIndex' => ['degrees' => 41.0, 'unit' => 'CELSIUS'],
        ],
        'hourlyForecast' => ['forecastHours' => []],
        'dailyForecast' => ['forecastDays' => []],
    ])->assertOk()->assertJson([
        'summary' => 'Hot and humid with a heat index of 41°C.',
        'healthAdvice' => 'Limit strenuous outdoor activity and stay hydrated.',
        'articleUrl' => 'https://www.cdc.gov/heat-health/about/index.html',
    ]);

    Http::assertSent(function ($request) {
        $body = $request->data();

        return $request->url() === 'https://openrouter.ai/api/v1/chat/completions'
            && $request->hasHeader('Authorization', 'Bearer test-openrouter-key')
            && $body['model'] === 'deepseek/deepseek-v4-flash-0731:online'
            && $body['response_format'] === ['type' => 'json_object']
            && $body['messages'][0]['role'] === 'system'
            && $body['messages'][1]['role'] === 'user'
            && str_contains($body['messages'][1]['content'], 'currentConditions');
    });
});

test('ai summary requires currentConditions', function (): void {
    $this->postJson('/api/ai/summary', [])
        ->assertStatus(400)
        ->assertJsonStructure(['message']);
});

test('ai summary returns 502 when OpenRouter is unreachable', function (): void {
    Http::fake(function () {
        throw new ConnectionException('Connection refused');
    });

    $this->postJson('/api/ai/summary', [
        'currentConditions' => ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
    ])->assertStatus(502)->assertJson([
        'message' => 'AI service is currently unreachable. Please try again later.',
    ]);
});

test('ai summary returns 502 when the API key is missing', function (): void {
    config()->set('services.openrouter.api_key', '');

    $this->postJson('/api/ai/summary', [
        'currentConditions' => ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
    ])->assertStatus(502)->assertJson([
        'message' => 'AI API key is not configured.',
    ]);
});

test('ai summary returns 502 when the model output is not JSON', function (): void {
    Http::fake([
        'openrouter.ai/*' => Http::response([
            'choices' => [['message' => ['content' => 'not json']]],
        ], 200),
    ]);

    $this->postJson('/api/ai/summary', [
        'currentConditions' => ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
    ])->assertStatus(502)->assertJson([
        'message' => 'AI service returned an invalid response.',
    ]);
});

test('ai summary returns 502 when a required field is missing', function (): void {
    Http::fake([
        'openrouter.ai/*' => Http::response([
            'choices' => [['message' => ['content' => json_encode([
                'summary' => 'Hot.',
                'healthAdvice' => 'Stay hydrated.',
            ])]]],
        ], 200),
    ]);

    $this->postJson('/api/ai/summary', [
        'currentConditions' => ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
    ])->assertStatus(502)->assertJson([
        'message' => 'AI service returned an invalid response.',
    ]);
});

test('ai summary returns 502 when the article URL is invalid', function (): void {
    Http::fake([
        'openrouter.ai/*' => Http::response([
            'choices' => [['message' => ['content' => json_encode([
                'summary' => 'Hot.',
                'healthAdvice' => 'Stay hydrated.',
                'articleUrl' => 'not-a-url',
            ])]]],
        ], 200),
    ]);

    $this->postJson('/api/ai/summary', [
        'currentConditions' => ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
    ])->assertStatus(502)->assertJson([
        'message' => 'AI service returned an invalid response.',
    ]);
});

test('service throws OpenRouterApiException on upstream error status', function (): void {
    Http::fake([
        'openrouter.ai/*' => Http::response(['error' => 'boom'], 500),
    ]);

    expect(fn () => app(OpenRouterService::class)->summarize(
        ['temperature' => ['degrees' => 33.0, 'unit' => 'CELSIUS']],
        [],
        [],
    ))->toThrow(OpenRouterApiException::class);
});